src/hnsw.c

```diff
@@ -119,15 +121,53 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
 	HnswGetMetaPageInfo(index, &m, NULL);
 	index_close(index, NoLock);
 
-	/* Approximate entry level */
-	entryLevel = (int) -log(1.0 / path->indexinfo->tuples) * HnswGetMl(m);
+	/*
+	 * HNSW cost estimation follows a formula that accounts for the total
+	 * number of tuples indexed combined with the parameters that most influence
+	 * the duration of the index scan, namely:
+	 *   m - the number of tuples that are scanned in each step of the HNSW
+	 *       graph traversal
+	 *   ef_search - which influences the total number of steps taken at layer 0
+	 *
+	 *  The source of the vector data can impact how many steps it takes to
+	 *  converge on the set of vectors to return to the executor. Currently,
+	 *  we use a hardcoded scaling factor (HNSWScanScalingFactor) to help
+	 *  influence that, but this could later become a configurable parameter
+	 *  based on the cost estimations.
+	 *
+	 * The tuple estimator formula is below:
+	 *
+	 * numIndexTuples = (totalLayers * m) +
+	 * 					(layer0TuplesMax * layer0Selectivity * HnswSF)
+	 *
+	 * "totalLayers * m" represents the floor of tuples we need to scan to get
+	 * to layer 0 (L0).
+	 *
+	 * "layer0TuplesMax" is the estimated total number of tuples we'd scan at
+	 * L0 if we weren't discarding already visited tuples as part of the scan.
+	 *
+	 * "layer0Selectivity" estimates the percentage of tuples that are scanned
+	 * at L0, accounting for previously visited tuples.
+	 *
+	 * "HnswSF" is a scaling factor, hardcoded to HnswScanScalingFactor
+	 */
+	totalLayers = floor(log(path->indexinfo->tuples) / log(m));
+	layer0TuplesMax = (double) (HnswGetLayerM(m, 0) * hnsw_ef_search);
+	layer0Selectivity = log(path->indexinfo->tuples) / (log(m) * (1 + log(hnsw_ef_search)));
 
-	/* TODO Improve estimate of visited tuples (currently underestimates) */
-	/* Account for number of tuples (or entry level), m, and ef_search */
-	costs.numIndexTuples = (entryLevel + 2) * m;
+	costs.numIndexTuples = (totalLayers * m) +
+		(HnswScanScalingFactor * layer0TuplesMax * layer0Selectivity);
 
 	genericcostestimate(root, path, loop_count, &costs);
 
+	/*
+	 * Explicitly set overall index selectivity after populating cost estimates,
+	 * as the generic cost estimator only factors in predicates as part of the
+	 * elimination strategy, whereas HNSW selecitivty is based on scanning the
+	 * index itself.
+	 */
+	costs.indexSelectivity = costs.numIndexTuples / path->indexinfo->tuples;
```

Comment:
Removing this should fix the test. I don't think this type of selectivity is what the planner expects.

I think we should do the same for IVFFlat ([ivfflat-cost branch](https://github.com/pgvector/pgvector/compare/ivfflat-cost)), which fixes two of the tests.

Reply:
Going to try that; if it works will push up with next commit.